File: 17.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type pos struct{ x, y int64 }
type key struct {
	ijet, irock int
	depths      [7]int
}

var (
	shapes = [][]string{
		{"####"},
		{".#.", "###", ".#."},
		{"..#", "..#", "###"},
		{"#", "#", "#", "#"},
		{"##", "##"},
	}
	directions = map[byte]int{'<': -1, '>': 1}
	states     map[key]struct{ rocks, height int64 }
	stones     map[pos]bool
	jets       string
	ijet       int
	maxheight  int64
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	jets = scanner.Text()
	fmt.Println(height(2022), height(1000000000000))
}

func drop(rock []string) {
	pp := pos{maxheight + 3 + int64(len(rock)), 2}
	for {
		dy := directions[jets[ijet%len(jets)]]
		ijet++
		hitStone := false
		for x, row := range rock {
			for y, ch := range row {
				if ch == '#' {
					if stones[pos{pp.x - int64(x), int64(y) + pp.y + int64(dy)}] {
						hitStone = true
						break
					}
				}
			}
		}
		if (dy > 0 && pp.y+int64(len(rock[0])) < 7 || dy < 0 && pp.y > 0) && !hitStone {
			pp.y += int64(dy)
		}
		if pp.x == 0 {
			break
		}
		if int64(len(rock))-pp.x-1 == 0 {
			break
		}
		pp.x--
		overlap := false
		for x, row := range rock {
			for y, ch := range row {
				if ch == '#' {
					if stones[pos{pp.x - int64(x), int64(y) + pp.y}] {
						overlap = true
						break
					}
				}
			}
		}
		pp.x++
		if overlap {
			break
		}
		pp.x--
	}
	for x, row := range rock {
		for y, ch := range row {
			if ch == '#' {
				_pp := pos{pp.x - int64(x), int64(y) + pp.y}
				stones[_pp] = true
				maxheight = max(maxheight, _pp.x)
			}
		}
	}
}

func height(threshold int64) int64 {
	maxheight = int64(-1)
	skipRocks, skipHeight := int64(0), int64(0)
	states = make(map[key]struct{ rocks, height int64 })
	stones = make(map[pos]bool)
	ijet = 0
	for r := int64(0); r+skipRocks < threshold; r++ {
		drop(shapes[r%int64(len(shapes))])
		depths := [7]int{-100, -100, -100, -100, -100, -100, -100}
		for y := 0; y < 7; y++ {
			for x := 0; x > -100; x-- {
				if stones[pos{maxheight + int64(x), int64(y)}] {
					depths[y] = x
					break
				}
			}
		}
		k := key{ijet % len(jets), int(r % int64(len(shapes))), depths}
		if prev, ok := states[k]; ok {
			idiff := r - prev.rocks
			remaining := threshold - r
			hdiff := maxheight - prev.height
			cycles := remaining / idiff
			states = make(map[key]struct{ rocks, height int64 })
			skipRocks, skipHeight = cycles*idiff, cycles*hdiff
		} else {
			states[k] = struct{ rocks, height int64 }{r, maxheight}
		}
	}
	return maxheight + 1 + skipHeight
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

File: 17.swift
import Foundation

struct Point: Hashable { var row, column: Int }

let shapes: [[String]] = [
    ["####"],
    [".#.",
     "###",
     ".#."],
    ["..#",
     "..#",
     "###"],
    ["#",
     "#",
     "#",
     "#"],
    ["##",
     "##"]
]
let directions: [Character: Int] = ["<": -1, ">": 1]
let jets = (readLine() ?? "").map{ directions[$0]! }

var nrows = 0
var ijet = 0
var stones = Set<Point>()

func overlap(_ rock: [String], _ p: Point) -> Bool {
    for row in 0..<rock.count {
        for (column, ch) in rock[row].enumerated() {
            if ch == "#" && stones.contains(Point(row: p.row - row, column: column + p.column)) {
                return true
            }
        }
    }
    return false
}

func drop(_ rock: [String]) {
    var p = Point(row: nrows + 2 + rock.count, column: 2)
    repeat {
        let jet = jets[ijet % jets.count]
        ijet += 1
        if (jet > 0 && p.column + rock[0].count < 7 ||
            jet < 0 && p.column > 0) &&
           !overlap(rock, Point(row: p.row, column: p.column + jet)) {
            p.column += jet
        }
        p.row -= 1
    } while p.row >= 0 && p.row + 1 - rock.count >= 0 && !overlap(rock, p)
    if p.row < 0 || p.row + 1 - rock.count < 0 || overlap(rock, p)  {
        p.row += 1
    }
    for row in 0..<rock.count {
        for (column, char) in rock[row].enumerated() {
            if char == "#" {
                stones.insert(Point(row: p.row - row, column: column + p.column))
            }
        }
        nrows = max(nrows, p.row - row + 1)
    }
}

func height(_ threshold: Int64) -> Int64 {
    struct Key: Hashable { var ijet, irock: Int; var depths: [Int] }
    var states: [Key: [Int64]] = [:]
    nrows = 0
    ijet = 0
    stones.removeAll()
    var r = Int64(), skipRocks = Int64(), skipHeight = Int64()
    while r + skipRocks < threshold {
        drop(shapes[Int(r % Int64(shapes.count))])
        var depths = [-100, -100, -100, -100, -100, -100, -100]
        for column in 0..<7 {
            for row in stride(from: nrows, to: nrows - 100, by: -1) {
                if (stones.contains(Point(row: row, column: column))) {
                    depths[column] = max(depths[column], nrows - row)
                }
            }
        }
        let key = Key(ijet: ijet % jets.count, irock: Int(r % Int64(shapes.count)), depths: depths)
        if let prev = states[key] {
            let length = r - prev[0]
            let cycles = (threshold - r) / length
            skipRocks = cycles * length
            skipHeight = cycles * (Int64(nrows) - prev[1])
            states.removeAll()
        } else {
            states[key] = [r, Int64(nrows)]
        }
        r += 1
    }
    return Int64(nrows) + skipHeight
}
print(height(2022), height(1000000000000))
